### DISCUSSION TITLE
Is it possible to have something like 'saved searches'

### DISCUSSION THREAD

**Discussion by apc** [emacs-citar/citar#116] [Q&A] [answered]
At the moment, calling `M-x bibtex-actions-open-pdf` gives a list of all entries in `bibtex-completion-bibliography`. But it seems as if it should be possible to only see those entries that have an associated PDF file (the same applies to `bibtex-actions-open-notes`). I do not yet understand how `completing-read-multiple` works to see if this would be easy to do (to list only those entries that match a particular search, e.g.), but in case anyone knows of a way to accomplish this, I'd love to hear about it!

**Comment by bdarcus** [comment]
On your specific examples, I added some of the details currently in bibtex-actions to make this easier.

So, for example, in my `orderless`-based setup, I can type this to give me all entries with notes, but no link:

![Screenshot from 2021-04-27 17-30-46](https://user-images.githubusercontent.com/1134/116315363-69f03580-a77e-11eb-98ee-0184ac9ce96d.png)

The substrings that are actually matched against are things like `has:pdf`.

Were you aware you can do that? Do you still think this would be valuable?

You're basically asking for pre-filtering, to obviate that.

I _think_ all that would be required is to add an optional INITIAL_INPUT parameter to the main `bibtex-actions-read` function, and to feed those strings to that on the appropriate "open" commands.

If you're asking to also allow arbitrary saved search strings, I'm not sure. I forget: does `ivy-bibtex` and/or `helm-bibtex` have that? If yes, how does it work there?

PS - the `consult--read` function, which is a nicer wrapper around `completing-read`, is interesting to look at, and easier to understand. I'm also curious about "history" though have never used it.

https://github.com/minad/consult/blob/ded314c4a3ee48e812b206016501cb49b29259fc/consult.el#L1654-1677

**Reply by apc** [reply]
> Were you aware you can do that? Do you still think this would be valuable?

Yes I was, though `:p` doesn't quite do what I want, since titles with colons and the letter `p` in them end up showing up (I'm using `prescient`; there's probably a way to make `:p` only match `has:pdf`, but I haven't really looked). At the moment, I call e.g. `bibtex-actions-open-pdf`, then type `:pdf`, etc. 

I think it would be valuable to have the pre-filtering in place, especially since it only makes sense to run `bibtex-actions-open-pdf` with entries with an associated PDF, no?

> 
> I _think_ all that would be required is to add an optional INITIAL_INPUT parameter to the main `bibtex-actions-read` function, and to feed those strings to that on the appropriate "open" commands.

OK, that's helpful. Would it make sense to modify the definition of `bibtex-actions-open-pdf`, say? 

> 
> If you're asking to also allow arbitrary saved search strings, I'm not sure. I forget: does `ivy-bibtex` and/or `helm-bibtex` have that? If yes, how does it work there?

Arbitrary saved search strings would be nice, I guess, though that's not what I had in mind. I don't think `ivy-bibtex` has that, and never really used `helm-bibtex`.

> 
> PS - the `consult--read` function, which is a nicer wrapper around `completing-read`, is interesting to look at, and easier to understand. I'm also curious about "history" though have never used it.
> 
> https://github.com/minad/consult/blob/ded314c4a3ee48e812b206016501cb49b29259fc/consult.el#L1654-1677

OK, I'll have a look! Thanks for your prompt and helpful reply, as always!

**Reply by bdarcus** [reply]
> Would it make sense to modify the definition of `bibtex-actions-open-pdf`, say?

Yes. 

So if this is the current definition ...

``` elisp
(defun bibtex-actions-open-pdf (keys)
 "Open PDF associated with the KEYS.
If multiple PDFs are found, ask for the one to open using
‘completing-read’."
  (interactive (list (bibtex-actions-read)))
  (bibtex-completion-open-pdf keys))
```

... I _think_ we want to end up with something like this (note the second to last line):

``` elisp
(defun bibtex-actions-open-pdf (keys)
 "Open PDF associated with the KEYS.
If multiple PDFs are found, ask for the one to open using
‘completing-read’."
  (interactive (list (bibtex-actions-read "has:pdf")))
  (bibtex-completion-open-pdf keys))
```

So that "has:pdf" would be the `completing-read` `INITIAL_INPUT` parameter.

... or if we can imagine extending the read function further, we could change it to use `cl-defun` (as `consult--read` does) and have named parameters; so:

``` elisp
(defun bibtex-actions-open-pdf (keys)
 "Open PDF associated with the KEYS.
If multiple PDFs are found, ask for the one to open using
‘completing-read’."
  (interactive (list (bibtex-actions-read :initial "has:pdf")))
  (bibtex-completion-open-pdf keys))
```

I'm not sure this the best path, but it's at least straightforward.

I guess another option is to filter the candidates before they even get to `completing-read`.

**Reply by bdarcus** [reply]
Here's a [good post](https://mail.gnu.org/archive/html/help-gnu-emacs/2018-05/msg00061.html) the describes the alternatives, though I'm not clear on how to do the "predicate" option.

**Reply by apc** [reply]
I guess which way to go might depend on whether it makes sense to allow folks to change their mind mid-search, as it were. That's what I often do, e.g. start looking for a citation to insert but then see another reference that might be relevant and decide to open it first. So the straightforward `INITIAL_INPUT` option might seem like a simple and flexible solution: `C-a C-k` if you change your mind and want to look for an entry that does not have an associated pdf, or if you thought you had a pdf associated with an entry and you can't see it in the list of matches for `has:pdf`, etc. 

Of course, it might be possible for all I know to simply call a function mid-search that removes the initial filter---that seems to be what is suggested  [here](https://mail.gnu.org/archive/html/help-gnu-emacs/2018-05/msg00068.html), in the same thread you mentioned above. (That post also gives the impression that 'good practices' tell against initial inputs altogether, but I don't know what to make of that.)

**Reply by apc** [reply]
From what I can understand (which isn't much), using `completion-table-in-turn` might be the way to go. You'd first need to generate a completion table with all candidates with pdfs, and then pass that as well as `bibtex-actions---candidates-cache` as arguments to `completion-table-in-turn`. I don't know how to do the first thing (if you point me in the right direction I can give it a try), but this could be an interesting alternative. 

**EDIT**: OK, I'm way out of my depth here. How do I check whether the 'text' of an element of `bibtex-actions---candidates-cache` contains a certain string? `string-match` doesn’t work because of the text properties I presume?

**Reply by bdarcus** [reply]
I don't think that's right.

I think the easy approach is just to use the initial value, because it's just a string.

And then if that doesn't work, use predicate with a filter function.

That function would use something like the built-in `seq-filter` or this `dash` function:

https://github.com/magnars/dash.el#-keep-fn-list

Both of those should work on the cached data.

I will open a branch/draft PR that points you in what I think is the right direction.

Edit: see #118.

**Reply by bdarcus** [reply]
> `string-match` doesn’t work because of the text properties I presume?

On this, I think it does:

``` elisp
ELISP> (setq test-foo (propertize "test" 'face 'bar))
#("test" 0 4
  (face bar))

ELISP> (string-match "st" test-foo)
2 (#o2, #x2, ?\C-b)
ELISP> (string-match "sat" test-foo)
nil
```

**Reply by apc** [reply]
Hm. I guess then I do not understand how the predicate filtering works. What am I doing wrong here? 

```
(defun my/has-pdf (string)
  (if (not (string-match "has:pdf" string))
      nil
      t))

(all-completions "Fodor" bibtex-actions--candidates-cache #'my/has-pdf)
```

I get `(wrong-type-argument stringp (#("Fodor ...` as an error. 

The documentation for `completion-table-with-predicate` says that the predicate arguments should be "a function of one argument which returns non-nil if and only if the argument is an element of TABLE which should be considered for completion."

If I call `my/has-pdf` with the `car` of `(all-completions "Fodor" bibtex-actions--candidates-cache)` as argument, I get a non-nil result (since that first entry has an associated pdf). So I don't know what I'm doing wrong.

**Reply by bdarcus** [reply]
> The documentation for `completion-table-with-predicate` says that the predicate arguments should be "a function of one argument" ...

I think this tripped me up earlier. For functions like affixation and such, I had to add an explicit argument to take the list of candidates, even though they weren't explicitly called from `completing-read-multiple`.

Something similar might be at issue here.

I pushed a commit just now for that filter function, but not sure this will work.

**Comment by bdarcus** [comment, marked as answer]
Not sure who's following this forum at this point, but just a head's up I merged support for this and for "preset" search strings.

See current README for details. Thanks to @apc for the idea, and the help.
